lifecycle-operator/controllers/common/schedulinggateshandler.go

```diff
@@ -0,0 +1,108 @@
+package common
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	klcv1alpha3 "github.com/keptn/lifecycle-toolkit/lifecycle-operator/apis/lifecycle/v1alpha3"
+	apicommon "github.com/keptn/lifecycle-toolkit/lifecycle-operator/apis/lifecycle/v1alpha3/common"
+	controllererrors "github.com/keptn/lifecycle-toolkit/lifecycle-operator/controllers/errors"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+//go:generate moq -pkg fake -skip-ensure -out ./fake/schedulinggateshandler_mock.go . ISchedulingGatesHandler
+type ISchedulingGatesHandler interface {
+	RemoveGates(ctx context.Context, workloadInstance *klcv1alpha3.KeptnWorkloadInstance) error
+	IsSchedulingEnabled() bool
+}
+
+type RemoveGatesFunc func(ctx context.Context, c client.Client, podName string, podNamespace string) error
+type GetPodsFunc func(ctx context.Context, c client.Client, ownerUID types.UID, ownerKind string, namespace string) ([]string, error)
+
+type SchedulingGatesHandler struct {
+	client.Client
+	logr.Logger
+	enabled     bool
+	removeGates RemoveGatesFunc
+	getPods     GetPodsFunc
+}
+
+func NewSchedulingGatesHandler(c client.Client, l logr.Logger, enabled bool) *SchedulingGatesHandler {
+	return &SchedulingGatesHandler{
+		Client:      c,
+		Logger:      l,
+		enabled:     enabled,
+		removeGates: removePodGates,
+		getPods:     getPodsOfOwner,
+	}
+}
+
+func (h *SchedulingGatesHandler) RemoveGates(ctx context.Context, workloadInstance *klcv1alpha3.KeptnWorkloadInstance) error {
+	switch workloadInstance.Spec.ResourceReference.Kind {
+	case "Pod":
+		return h.removeGates(ctx, h.Client, workloadInstance.Spec.ResourceReference.Name, workloadInstance.Namespace)
+	case "ReplicaSet", "StatefulSet", "DaemonSet":
+		podList, err := h.getPods(ctx, h.Client, workloadInstance.Spec.ResourceReference.UID, workloadInstance.Spec.ResourceReference.Kind, workloadInstance.Namespace)
+		if err != nil {
+			h.Logger.Error(err, "cannot get pods")
+			return err
+		}
+		for _, pod := range podList {
+			err := h.removeGates(ctx, h.Client, pod, workloadInstance.Namespace)
+			if err != nil {
+				h.Logger.Error(err, "cannot remove gates from pod")
+				return err
+			}
+		}
+	default:
+		return controllererrors.ErrUnsupportedWorkloadInstanceResourceReference
+	}
+
+	return nil
+}
+
+func (h *SchedulingGatesHandler) IsSchedulingEnabled() bool {
+	return h.enabled
+}
+
+func removePodGates(ctx context.Context, c client.Client, podName string, podNamespace string) error {
+	pod := &v1.Pod{}
+	err := c.Get(ctx, types.NamespacedName{Namespace: podNamespace, Name: podName}, pod)
+	if err != nil {
+		return err
+	}
+
+	if pod.Annotations[apicommon.SchedulingGateRemoved] != "" {
+		return nil
+	}
+
+	if len(pod.Annotations) == 0 {
+		pod.Annotations = make(map[string]string, 1)
+	}
+	pod.Annotations[apicommon.SchedulingGateRemoved] = "true"
+	pod.Spec.SchedulingGates = nil
```

Comment:
why we need annotations? I would only rely on the scheduling gate to be removed from the spec 🤔 
In the webhook, we would need two webhooks: 
- on create that adds the gate
- on update that checks if the gate was removed to allow or not the pod

Reply:
We discussed this in the team and agreed that it is beneficial to have annotation about removal of the gates in place. There reason is that we can easily check that the scheduling gates were added and afterwards removed and the pod didn't pass through without the expected blocking. 

to the 2 mutating webhooks, I agree, but we all agreed on the team that this can be a follow up, as the webhook logic needs refactoring and polishing (out of scope for this), here is a ticket prepared https://github.com/keptn/lifecycle-toolkit/issues/2153